File: apps/pwa/src/server/api/get_profile.ts
import { prefixServerOrigin } from '@/global_states/server';
import { HEADER_TOKEN } from '@/constants/api';
import { request } from '..';

interface Response {
  id: string;
  username: string;
  avatar: string;
  nickname: string;
  joinTimestamp: number;
  admin: 0 | 1;
  musicbillOrdersJSON?: string | null;
  lastActiveTimestamp: number;
  twoFAEnabled: boolean;
}

async function getProfile(token?: string) {
  const profile = await request<Response>({
    path: '/api/profile',
    headers: token
      ? {
          [HEADER_TOKEN]: token,
        }
      : undefined,
    withToken: !token,
  });
  return {
    ...profile,
    avatar: prefixServerOrigin(profile.avatar),
  };
}

export default getProfile;

File: apps/flutter/lib/server/api/get_profile.dart
import 'package:cicada/utils/prefix_server_origin.dart';

import '../../constants/index.dart';
import '../../server/request.dart';

class Profile {
  final String id;
  final String username;
  final String? avatar;
  final String nickname;
  final int joinTimestamp;
  final bool twoFAEnabled;

  Profile({
    required this.id,
    required this.username,
    required this.avatar,
    required this.nickname,
    required this.joinTimestamp,
    required this.twoFAEnabled,
  });

  factory Profile.fromJSON(Map<String, dynamic> json) => Profile(
    id: json['id'],
    username: json['username'],
    avatar: prefixServerOrigin(json['avatar']),
    nickname: json['nickname'],
    joinTimestamp: json['joinTimestamp'],
    twoFAEnabled: json['twoFAEnabled'],
  );
}

Future<Profile> getProfile(String? token) async {
  Map<String, String> headers = {};
  if (token != null) {
    headers[TOKEN_HEADER_KEY] = token;
  }
  final responseData = await httpGet(path: "/api/profile", headers: headers);
  return Profile.fromJSON(responseData);
}
